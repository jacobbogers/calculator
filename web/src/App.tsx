/* vender */
import React, { useCallback, useMemo } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import classNames from 'classnames';

/* styles */
import styles from './App.module.scss';

/* components */
import CalculatorBtn from './components/CalculatorBtn';
import KeyBoard from './components/KeyBoard';
import Display from './components/Display';

/* redux */
import type { StoreLayout } from './lib/reducers';

/* misc */
import { createAction } from './lib/actions';

export default function App() {
    const dispatch = useDispatch();
    const isDirty = useSelector<{ tainted: boolean }, boolean>((state) => state.tainted);
    const networkRequestInTransit = useSelector<StoreLayout, boolean>((state) => state.inTransit);

    const safeDispatch = useCallback(
        (...args: Parameters<typeof dispatch>) => {
            if (networkRequestInTransit) {
                return;
            }
            return dispatch(...args);
        },
        [dispatch, networkRequestInTransit]
    );

    const className = useMemo(
        () =>
            classNames({
                [styles.app]: true,
                [styles.wait]: networkRequestInTransit
            }),
        [networkRequestInTransit]
    );

    return (
        <div className={className}>
            <KeyBoard>
                <Display />
                {/* row 1 */}
                <CalculatorBtn disabled darker>
                    Rad
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    Deg
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    x!
                </CalculatorBtn>

                <CalculatorBtn disabled darker>
                    (
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    )
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    %
                </CalculatorBtn>
                {isDirty ? (
                    <CalculatorBtn
                        data-testid="btnClearEntry"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearEntry'))}
                    >
                        CE
                    </CalculatorBtn>
                ) : (
                    <CalculatorBtn
                        data-testid="btnClearAll"
                        darker
                        smaller
                        onClick={() => safeDispatch(createAction('clearAll'))}
                    >
                        AC
                    </CalculatorBtn>
                )}
                {/* row 2 */}

                <CalculatorBtn disabled darker>
                    Inv
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    sin
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    ln
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn7" onClick={() => safeDispatch(createAction('digit', '7'))}>
                    7
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn8" onClick={() => safeDispatch(createAction('digit', '8'))}>
                    8
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn9" onClick={() => safeDispatch(createAction('digit', '9'))}>
                    9
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnDivide"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('divide'))}
                >
                    ÷
                </CalculatorBtn>

                {/* row 3 */}

                <CalculatorBtn disabled darker>
                    π
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    cos
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    log
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn4" onClick={() => safeDispatch(createAction('digit', '4'))}>
                    4
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn5" onClick={() => safeDispatch(createAction('digit', '5'))}>
                    5
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn6" onClick={() => safeDispatch(createAction('digit', '6'))}>
                    6
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnMultiply"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('multiply'))}
                >
                    ×
                </CalculatorBtn>

                {/* row 4 */}

                <CalculatorBtn disabled darker>
                    e
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    tan
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    √
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn1" onClick={() => safeDispatch(createAction('digit', '1'))}>
                    1
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn2" onClick={() => safeDispatch(createAction('digit', '2'))}>
                    2
                </CalculatorBtn>
                <CalculatorBtn data-testid="btn3" onClick={() => safeDispatch(createAction('digit', '3'))}>
                    3
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubtract"
                    darker
                    larger
                    onClick={() => safeDispatch(createAction('subtract'))}
                >
                    −
                </CalculatorBtn>

                {/* row 5 */}

                <CalculatorBtn disabled darker>
                    Ans
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    EXP
                </CalculatorBtn>
                <CalculatorBtn disabled darker>
                    <span>xʸ</span>
                </CalculatorBtn>

                <CalculatorBtn data-testid="btn0" onClick={() => safeDispatch(createAction('digit', '0'))}>
                    0
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnDot" bolder onClick={() => safeDispatch(createAction('digit', '.'))}>
                    .
                </CalculatorBtn>
                <CalculatorBtn
                    data-testid="btnSubmit"
                    bolder
                    larger
                    submit
                    onClick={() => safeDispatch(createAction('result'))}
                >
                    =
                </CalculatorBtn>
                <CalculatorBtn data-testid="btnAdd" darker larger onClick={() => safeDispatch(createAction('add'))}>
                    +
                </CalculatorBtn>
            </KeyBoard>
        </div>
    );
}
