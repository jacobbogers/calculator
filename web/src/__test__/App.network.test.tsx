import React from 'react';
import type { Store } from 'redux';
import type { MockResponseInit } from 'vitest-fetch-mock';
import { cleanup, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import createStore from '../lib/store';
import App from '../App';

import { getUIElements, customRender } from './test-helper';
import type { StoreLayout } from '../lib/reducers';

describe('Network fetch', () => {
    let store: Store<StoreLayout>;

    beforeEach(() => {
        store = createStore();
        fetchMock.enableMocks();
    });

    afterEach(() => {
        cleanup();
        fetchMock.resetMocks();
    });
    it('Non http-ok response, Non JSON response', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { lastHistoryLine, commandLine, btn4, btnSubtract, btnDot, btnSubmit } = getUIElements(screen);
        await user.click(btn4);
        await user.click(btnSubtract);
        await user.click(btnDot);
        await user.click(btn4);

        expect(lastHistoryLine.textContent).toBe('Ans = 0');
        expect(commandLine.textContent).toBe('4 - .4');
        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'text/html'
                    },
                    status: 500,
                    statusText: 'Not Found'
                },
                body: '<html><head></head><body>Error</body></head>'
            });
        });

        await user.click(btnSubmit);
        expect(lastHistoryLine.textContent).toBe('4 - .4 =');
        expect(commandLine.textContent).toBe('Error');
        await user.click(btnSubmit);
        expect(store.getState().serverError).toMatchSnapshot();
        unmount();
    });
    it('Non http-ok response, it is JSON response, but rejected by schema', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { lastHistoryLine, commandLine, btn4, btnSubtract, btnDot, btnSubmit } = getUIElements(screen);
        await user.click(btn4);
        await user.click(btnSubtract);
        await user.click(btnDot);
        await user.click(btn4);

        expect(lastHistoryLine.textContent).toBe('Ans = 0');
        expect(commandLine.textContent).toBe('4 - .4');
        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    },
                    status: 400,
                    statusText: 'Not Found'
                },
                body: '{ "hello":"world" }'
            });
        });

        await user.click(btnSubmit);
        expect(lastHistoryLine.textContent).toBe('4 - .4 =');
        expect(commandLine.textContent).toBe('Error');
        await user.click(btnSubmit);
        expect(store.getState().serverError).toMatchSnapshot();

        unmount();
    });
    it('http-ok response, but non JSON response', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { lastHistoryLine, commandLine, btn4, btnSubtract, btnDot, btnSubmit } = getUIElements(screen);
        await user.click(btn4);
        await user.click(btnSubtract);
        await user.click(btnDot);
        await user.click(btn4);

        expect(lastHistoryLine.textContent).toBe('Ans = 0');
        expect(commandLine.textContent).toBe('4 - .4');
        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'text/html'
                    },
                    status: 200,
                    statusText: 'Ok'
                },
                body: '<html><head></head><body>Welcome to nginx</body></head>'
            });
        });

        await user.click(btnSubmit);
        expect(lastHistoryLine.textContent).toBe('4 - .4 =');
        expect(commandLine.textContent).toBe('Error');
        await user.click(btnSubmit);
        expect(store.getState().serverError).toMatchSnapshot();

        unmount();
    });
    it('http-ok response, but JSONschema rejects response', async () => {
        const { unmount } = customRender(<App />, store);
        const user = userEvent.setup({ delay: 0.4 });
        const { lastHistoryLine, commandLine, btn4, btnSubtract, btnDot, btnSubmit } = getUIElements(screen);
        await user.click(btn4);
        await user.click(btnSubtract);
        await user.click(btnDot);
        await user.click(btn4);

        expect(lastHistoryLine.textContent).toBe('Ans = 0');
        expect(commandLine.textContent).toBe('4 - .4');
        await fetchMock.mockResponse((request: Request) => {
            return Promise.resolve<MockResponseInit>({
                init: {
                    headers: {
                        'content-type': 'application/json'
                    },
                    status: 200,
                    statusText: 'Ok'
                },
                body: '{ "schema":"reject" }'
            });
        });

        await user.click(btnSubmit);
        expect(lastHistoryLine.textContent).toBe('4 - .4 =');
        expect(commandLine.textContent).toBe('Error');
        await user.click(btnSubmit);
        expect(store.getState().serverError).toMatchSnapshot();

        unmount();
    });
});
