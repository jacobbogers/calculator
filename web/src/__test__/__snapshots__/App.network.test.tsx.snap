// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Network fetch > Non http-ok response, Non JSON response 1`] = `
{
  "error": {
    "code": 9300,
    "message": "Server responded with an error. The response was not in json format: <html><head></head><body>Error</body></head>",
    "reqId": "N/A",
  },
}
`;

exports[`Network fetch > Non http-ok response, it is JSON response, but rejected by schema 1`] = `
{
  "error": {
    "code": 9400,
    "message": "Server responded with an error. The response was in json, but unknown format: { \\"hello\\":\\"world\\" }",
    "reqId": "N/A",
  },
}
`;

exports[`Network fetch > http-ok response, but JSONschema rejects response 1`] = `
{
  "error": {
    "code": 9600,
    "message": "Server responded ok. THe response was in json, but unknown format: { \\"schema\\":\\"reject\\" }",
    "reqId": "N/A",
  },
}
`;

exports[`Network fetch > http-ok response, but non JSON response 1`] = `
{
  "error": {
    "code": 9500,
    "message": "Server responded ok. The response was not in json format: <html><head></head><body>Welcome to nginx</body></head>",
    "reqId": "N/A",
  },
}
`;
